src/test/ui/consts/array-literal-index-oob.stderr

```diff
@@ -12,7 +12,7 @@ error: reaching this expression at runtime will panic or abort
 LL |     &{[1, 2, 3][4]};
    |     --^^^^^^^^^^^^-
    |       |
-   |       index out of bounds: the len is 3 but the index is 4
+   |       indexing out of bounds: the len is 3 but the index is 4
```

Comment:
So curiously something is printing the `UndefinedBehaviorInfo` message here -- the source of this error is a `throw_ub!`, *not* a `throw_panic!`. And the header "reaching this expression at runtime will panic or abort" indicates that that something is ConstEval. But ConstEval should never hit UB here, because that UB is guarded by a bounds check and an `Assert` MIR terminator!

I have no idea how that is possibe. Cc https://github.com/rust-lang/rust/issues/61821

Reply:
Something *very* strange is going on... this is the generated MIR for the promoted:
```
promoted[0] in  main: i32 = {
    let mut _0: i32;                     // return place in scope 0 at src/main.rs:3:5: 3:20
    let mut _1: i32;                     // in scope 0 at src/main.rs:3:6: 3:20
    let mut _2: [i32; 3];                // in scope 0 at src/main.rs:3:7: 3:16
    let mut _3: usize;                   // in scope 0 at src/main.rs:3:17: 3:18

    bb0: {
        _2 = [const 1i32, const 2i32, const 3i32]; // bb0[0]: scope 0 at src/main.rs:3:7: 3:16
        _3 = const 4usize;               // bb0[1]: scope 0 at src/main.rs:3:17: 3:18
        _1 = _2[_3];                     // bb0[2]: scope 0 at src/main.rs:3:7: 3:19
        _0 = move _1;                    // bb0[3]: scope 0 at src/main.rs:3:5: 3:20
        return;                          // bb0[4]: scope 0 at src/main.rs:3:5: 3:20
    }
}
```
Notice the missing bounds check! Somehow, when promoting, we create MIR that would never actually be created for a normal function...??

Reply:
Cc @ecstatic-morse @eddyb  for promotion knowledge. See above, any idea what might be happening here?

Reply:
Promoteds don't have bounds checks because the bounds checks are still in the runtime code to make sure panics don't disappear or become compile time errors

Reply:
We could silence UB errors in promoteds because the only way they van happen is if the runtime will panic before accessing the promoted

Reply:
> Promoteds don't have bounds checks because the bounds checks are still in the runtime code to make sure panics don't disappear or become compile time errors

I thought they would get duplicated then, so that the MIR extracted into the promoted is still type-safe.

> This is exactly what the error is telling us, so... all good?

Well currently we show two errors: one from running the promoted, and one (I think) from ConstProp of the `Assert` that did not get promoted.

Things don't have to be cleaned up by this PR but I was wondering what is going on.

Reply:
But anyway this is pre-existing, so probably https://github.com/rust-lang/rust/issues/61821 is the better place for discussion. Sorry for starting it here.